src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -0,0 +1,472 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Analyzer.Utilities.FlowAnalysis.Analysis.TaintedDataAnalysis
+{
+    internal partial class TaintedDataAnalysis
+    {
+        private sealed class TaintedDataOperationVisitor : AnalysisEntityDataFlowOperationVisitor<TaintedDataAnalysisData, TaintedDataAnalysisContext, TaintedDataAnalysisResult, TaintedDataAbstractValue>
+        {
+            /// <summary>
+            /// Mapping of a tainted data sinks to their originating sources.
+            /// </summary>
+            /// <remarks>Keys are <see cref="SymbolAccess"/> sinks where the tainted data entered, values are <see cref="SymbolAccess"/>s where the tainted data originated from.</remarks>
+            private Dictionary<SymbolAccess, HashSet<SymbolAccess>> TaintedSourcesBySink { get; set; }
+
+            public TaintedDataOperationVisitor(TaintedDataAnalysisContext analysisContext)
+                : base(analysisContext)
+            {
+                this.TaintedSourcesBySink = new Dictionary<SymbolAccess, HashSet<SymbolAccess>>();
+            }
+
+            public ImmutableArray<TaintedDataSourceSink> GetTaintedDataSourceSinkEntries()
+            {
+                ImmutableArray<TaintedDataSourceSink>.Builder builder = ImmutableArray.CreateBuilder<TaintedDataSourceSink>();
+                foreach (KeyValuePair<SymbolAccess, HashSet<SymbolAccess>> kvp in this.TaintedSourcesBySink)
+                {
+                    builder.Add(
+                        new TaintedDataSourceSink(
+                            kvp.Key,
+                            SinkKind.Sql,
+                            ImmutableArray.Create<SymbolAccess>(kvp.Value.ToArray())));
+                }
+                 
+                return builder.ToImmutableArray();
+            }
+
+            protected override void AddTrackedEntities(ImmutableArray<AnalysisEntity>.Builder builder)
+            {
+                this.CurrentAnalysisData.AddTrackedEntities(builder);
+            }
+
+            protected override bool Equals(TaintedDataAnalysisData value1, TaintedDataAnalysisData value2)
+            {
+                return value1.Equals(value2);
+            }
+
+            protected override TaintedDataAbstractValue GetAbstractDefaultValue(ITypeSymbol type)
+            {
+                return TaintedDataAbstractValue.NotTainted;
```

Comment:
>TaintedDataAbstractValue.NotTainted [](start = 23, length = 35)

You return `TaintedDataAbstractValue.NotTainted` here but `CoreTaintedDataAnalysisDataDomain.GetDefaultValue` returns `TaintedDataAbstractValue.Unknown`?

Reply:
Wondering if you can combine these two, such that everything is non-tainted by default, and can only change to tainted, so your enum can have just 2 values. Thoughts?

---
In reply to: [234015291](https://github.com/dotnet/roslyn-analyzers/pull/1882#discussion_r234015291) [](ancestors = 234015291)

Reply:
That's a good question.  :-)

NotTainted and Tainted are pretty much what I care about, and throwing Unknown into the mix clouds what's going on.  I think I'll give two values a shot.